src/main/xerus/monstercat/api/Player.kt

```diff
@@ -164,13 +173,48 @@ object Player: FadingHBox(true, targetHeight = 25) {
 		}
 	}
 	
-	private val pauseButton = ToggleButton().id("play-pause").onClick { if(isSelected) player?.pause() else player?.play() }
-	private val stopButton = buttonWithId("stop") { reset() }
-	private val volumeSlider = Slider(0.0, 1.0, Settings.PLAYERVOLUME()).scrollable(0.05).apply {
-		prefWidth = 100.0
-		valueProperty().listen { updateVolume() }
-	}
-	
+	private val pauseButton = ToggleButton().id("play-pause")
+			.tooltip("Pause / Play")
+			.onClick { if (isSelected) player?.pause() else player?.play() }
+	private val stopButton = Button().id("stop")
+			.tooltip("Stop playing")
+			.onClick {
+				reset()
+				Playlist.clear()
+			}
+	private val volumeSlider = Slider(0.0, 1.0, Settings.PLAYERVOLUME())
+			.scrollable(0.05)
+			.apply {
+				prefWidth = 100.0
+				valueProperty().listen { updateVolume() }
+				tooltip = Tooltip("Volume")
+			}
+	private val shuffleButton = ToggleButton().id("shuffle")
+			.tooltip("Shuffle")
+			.apply { selectedProperty().bindBidirectional(Playlist.shuffle) }
+	private val repeatButton = ToggleButton().id("repeat")
+			.tooltip("Repeat all")
+			.apply { selectedProperty().bindBidirectional(Playlist.repeat) }
+	private val skipbackButton = buttonWithId("skipback") { playPrev() }
+			.tooltip("Previous")
+			.apply { Playlist.currentIndex.addListener { _, _, newValue ->
+				val disable = (newValue == 0)
+				isDisable = disable
+				isVisible = !disable
+			} }
+	private val skipButton = buttonWithId("skip") { playNext() }
+			.tooltip("Next")
+			.apply {
+				fun disable(index: Int? = Playlist.currentIndex.value, repeat: Boolean? = Playlist.repeat.value, random: Boolean? = Playlist.shuffle.value) {
+					val disable = index == Playlist.tracks.lastIndex && repeat == false && random == false
+					isDisable = disable
+					isVisible = !disable
+				}
+				Playlist.currentIndex.addListener { _, _, newValue -> disable(index = newValue) }
+				Playlist.repeat.addListener { _, _, newValue -> disable(repeat = newValue)}
+				Playlist.shuffle.addListener { _, _, newValue -> disable(random = newValue)}
```

Comment:
what the heck is this piece of code doing?

And regardless of what it does, you could probably use `arrayOf(...).listen {` to listen to all properties at once.

Reply:
This is to disable the "next" button when we are on the last track of the playlist as you asked me to do. The thing is it should still be enabled if we have "shuffle" on, or "repeat" on.

Reply:
I can't use arrayOf().addListener because : 1. it doesn't give me the observable and 2. I need different checks for "currentIndex" (= an Int) and "shuffle"/"repeat" (= booleans)

Reply:
Update : 0e108aedc082ddbd57506638ac87f7b301a973b0

Reply:
It shouldn't work like this. Now it will still disable if you go to the lastIndex with repeat on. 
You need to always use all three values, but you don't need to pass the new value of the Listener to anything, you can just use the value of the observable.

Reply:
I'll try to find a working way. Or maybe just remove the feature ?

Reply:
No, it is super simple.

Reply:
```kotlin
arrayOf(Playlist.currentIndex, Playlist.repeat, Playlist.shuffle).listen {
	isDisable = index == Playlist.tracks.lastIndex && !Playlist.repeat.value && !Playlist.shuffle.value
}
```
that should roughly work

Reply:
.listen doesn't exist, it's addListener.

Reply:
5b6ae0ebbb71ee610e8474cd35891528677cac08